astroid/constraint.py

```diff
@@ -0,0 +1,135 @@
+# Licensed under the LGPL: https://www.gnu.org/licenses/old-licenses/lgpl-2.1.en.html
+# For details: https://github.com/PyCQA/astroid/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
+
+"""Classes representing different types of constraints on inference values."""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TypeVar, Union
+
+from astroid import nodes, util
+
+_NameNodes = Union[nodes.AssignAttr, nodes.Attribute, nodes.AssignName, nodes.Name]
+_ConstraintT = TypeVar("_ConstraintT", bound="Constraint")
+
+
+class Constraint(ABC):
+    """Represents a single constraint on a variable."""
+
+    def __init__(self, node: nodes.NodeNG, negate: bool) -> None:
+        self.node = node
+        """The node that this constraint applies to."""
+        self.negate = negate
+        """True if this constraint is negated. E.g., "is not" instead of "is"."""
+
+    @classmethod
+    @abstractmethod
+    def match(
+        cls: _ConstraintT, node: _NameNodes, expr: nodes.NodeNG, negate: bool = False
+    ) -> _ConstraintT | None:
+        """Return a new constraint for node matched from expr, if expr matches
+        the constraint pattern.
+
+        If negate is True, negate the constraint.
+        """
+
+    @abstractmethod
+    def satisfied_by(self, inferred: nodes.NodeNG | type[util.Uninferable]) -> bool:
```

Comment:
```suggestion
    def satisfied_by(self, inferred: InferenceResult) -> bool:
```
See https://github.com/PyCQA/astroid/pull/1189#discussion_r903134872

`InferenceResult` can be imported from `astroid/typing.py`.

Reply:
This was done at some point.